File: apps/web/src/features/categories/api/list-categories.ts
import { api } from '@/api/client'
import type { Category, CategoryKind } from '@/features/categories/category'

export async function listCategories(kind: CategoryKind, includeInactive = false): Promise<Category[]> {
  const { data, error } = await api.GET('/api/v1/categories', {
    params: { query: { kind, include_inactive: includeInactive } },
  })
  if (error) throw new Error('Failed to load categories')
  return data
}

File: apps/api/internal/categories/application/list_categories.go
package application

import (
	"context"

	"caudal-api/internal/categories/domain"
	"caudal-api/internal/categories/ports"
)

type ListCategories struct {
	repository ports.Repository
}

func NewListCategories(repository ports.Repository) ListCategories {
	return ListCategories{repository: repository}
}

type ListCategoriesQuery struct {
	Kind            domain.Kind
	IncludeInactive bool
}

func (useCase ListCategories) Execute(ctx context.Context, query ListCategoriesQuery) ([]domain.Category, error) {
	return useCase.repository.List(ctx, query.Kind, query.IncludeInactive)
}
